lib/sampling.h

```diff
@@ -8,36 +8,44 @@
 #include <tuple>
 
 namespace {
-    static constexpr float M_2PI = 2.f * M_PI;
+
+static constexpr float M_2PI = 2.f * M_PI;
+
 }
 
 namespace sampling {
 
-    xorshift64star<float> uniform{4};
+__thread static xorshift64star<float> uniform{4};
```

Comment:
Nice. Thanks! One last comment: lets put it also in anonymous namespace above. Then it can't be accessed from outside anymore, and this is exactly what we want.